lib/view_model/active_record/update_operation.rb

```diff
@@ -131,33 +129,36 @@ def run!(deserialize_context:)
         # deletes are handled elsewhere.
 
         changes = viewmodel.changes
-
         if changes.new? || changes.changed_attributes.present? || changes.changed_associations.present?
+          debug "-> #{debug_name}: Validating changes"
+          # invoke pre-save hook
           viewmodel.before_save(changes, deserialize_context: deserialize_context)
 
-          # The hook before this might have caused additional changes. All changes should be checked by the policy,
-          # so we have to recalculate the change set.
-
-          changes = viewmodel.changes
+          # Invoke any validation hooks before attempting to save.
+          viewmodel.validate!
 
-          deserialize_context.editable!(viewmodel,
-                                        initial_editability: initial_editability,
-                                        changes: changes)
+          # The hooks before this might have caused additional changes. All
+          # changes should be checked by the policy, so we have to recalculate
+          # the change set. Note that changes to associated children that point
+          # to this parent are recorded in `changes`, but their models have not
+          # yet been updated.
+          final_changes = viewmodel.changes
+          deserialize_context.run_callback(ViewModel::Callbacks::Hook::OnChange, viewmodel, changes: final_changes)
         end
 
-        # Invoke any validation hooks before attempting to save.
-        viewmodel.validate!
-
-        debug "-> #{debug_name}: Saving"
-        begin
-          model.save!
-        rescue ::ActiveRecord::RecordInvalid => ex
-          raise ViewModel::DeserializationError::Validation.from_active_model(ex.errors, blame_reference)
-        rescue ::ActiveRecord::StaleObjectError => ex
-          raise ViewModel::DeserializationError::LockFailure.new(blame_reference)
+        # Save if the model has been altered. Covers not only models with
+        # view changes but also lock version assertions.
```

Comment:
If the lock version causes `changed?` then we already know we've failed. I think the code is fine, but the lock version assertions with no other changes seem to not be supported in rails, unless also using explicit `.lock()`.

Reply:
Nothing to do here, but acknowledged that if `model.changes` includes `lock_version` we know already that it would fail due to lock assertion before going to the database. I don't think it's worth checking that explicitly though. Maybe PR to Rails to add that check in there instead.